contracts/examples/mintable-erc1155-transfer/FxMintableERC1155ChildTunnel.sol

```diff
@@ -0,0 +1,261 @@
+    bytes32 public constant WITHDRAW_BATCH = keccak256("WITHDRAW_BATCH");
+
+    event TokenMapped(address indexed rootToken, address indexed childToken);
+    event FxWithdrawMintableERC1155(
+        address indexed rootToken,
+        address indexed childToken,
+        address indexed userAddress,
+        uint256 id,
+        uint256 amount
+    );
+    event FxDepositMintableERC1155(
+        address indexed rootToken,
+        address indexed depositor,
+        address indexed userAddress,
+        uint256 id,
+        uint256 amount
+    );
+    event FxBatchWithdrawMintableERC1155(
+        address indexed rootToken,
+        address indexed childToken,
+        address indexed userAddress,
+        uint256[] ids,
+        uint256[] amounts
+    );
+    event FxBatchDepositMintableERC1155(
+        address indexed rootToken,
+        address indexed depositor,
+        address indexed userAddress,
+        uint256[] ids,
+        uint256[] amounts
+    );
+    // root to child token
+    mapping(address => address) public rootToChildToken;
+    // child token template
+    address public immutable childTokenTemplate;
+    // root token template codehash
+    bytes32 public immutable rootTokenTemplateCodeHash;
+
+    constructor(
+        address _fxChild,
+        address _childTokenTemplate,
+        address _rootTokenTemplate
+    ) FxBaseChildTunnel(_fxChild) {
+        childTokenTemplate = _childTokenTemplate;
+        require(
+            Address.isContract(_childTokenTemplate),
+            "FxMintableERC1155ChildTunnel: Token template is not contract"
+        );
+        // compute root token template code hash
+        rootTokenTemplateCodeHash = keccak256(minimalProxyCreationCode(_rootTokenTemplate));
+    }
+
+    //
+    // External methods
+    //
+
+    // deploy child token with unique id
+    function deployChildToken(uint256 _uniqueId, string calldata _uri) external {
+        // deploy new child token using unique id
+        address childToken = createClone(keccak256(abi.encodePacked(_uniqueId)), childTokenTemplate); // child salt, childTokenTemplate
+
+        // compute root token address before deployment using create2
+        address rootToken = computedCreate2Address(
+            keccak256(abi.encodePacked(childToken)), // root salt
+            rootTokenTemplateCodeHash,
+            fxRootTunnel
+        );
+
+        // check if mapping is already there
+        require(rootToChildToken[rootToken] == address(0x0), "FxMintableERC1155ChildTunnel: ALREADY_MAPPED");
+        rootToChildToken[rootToken] = childToken;
+        emit TokenMapped(rootToken, childToken);
+
+        // initialize child token with all parameters
+        FxERC1155(childToken).initialize(address(this), rootToken, _uri);
+    }
+
+    // Mint tokens on child chain
+    function mintToken(
+        address _childToken,
+        uint256 _tokenId,
+        uint256 _amount,
+        bytes calldata _data
+    ) external {
```

Comment:
Shouldn't the access of this function be restricted?

Reply:
Yes, but this is covered inside the function. Mint should not be allowed by anyone, only the deployer or it's use case specific. This check exists when calling `childTokenContract.mint(...)` method ie [FxERC1155 mint call](https://github.com/fx-portal/contracts/blob/main/contracts/tokens/FxERC1155.sol#L41).

Reply:
https://github.com/fx-portal/contracts/pull/41#discussion_r1018371415

Reply:
Yes, you're right. I was trying to justify @QEDK's [comment](https://github.com/fx-portal/contracts/pull/41/files/3614e29f5380b50e87d338656b261e299614fc0c#r1017266352) & had overlooked this! 
I'm not sure if it makes sense to let anyone deploy a mintable child token but only have the owner mint them. We can 1. do this or, 2. implement [OZ Roles](https://docs.openzeppelin.com/contracts/2.x/api/access#Roles): possibly Deployer & Minter and standardise them across other contracts (`FxMintableERC20ChildTunnel` & `FxMintableERC721ChildTunnel`), or 3. keep both these functions ristricted to onlyOwner as it was initially written and leave this functionality to added depending on the consumer dev's use case.

What do you think?

Reply:
Here's what I am thinking, move the ownership logic to a special mintable token, set ownership to the token mapper at time of deployment (which is left accessible), and give minting rights as well, it seems much cleaner, any scenario where message-passing isn't happening can happen on the token itself, thoughts? @dev1644

Reply:
7045744ff48ff7070d5c0288e08e089c7f6daa61